substrate/frame/treasury/src/lib.rs

```diff
@@ -1020,6 +1029,72 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			// Must never be less than 0 but better be safe.
 			.saturating_sub(T::Currency::minimum_balance())
 	}
+
+	/// Ensure the correctness of the state of this pallet.
+	///
+	/// ## Invariants
+	///
+	/// 1. [`ProposalCount`] >= [`Proposals`].count(). The [`ProposalCount`] SV is only increased, but never
+	///    decreased whereas individual proposals can be removed from the [`Proposals`] SM.
+	/// 2. [`T::ProposalBondMinimum`] * Number of proposals with non-zero bond <= sum_{p in Proposals} p.bond.
+	/// 3. (Optional) [`T::ProposalBondMaximum`] * Number of proposals with non-zero bond >= sum_{p in Proposals} p.bond.
+	/// 4. Each [`ProposalIndex`] contained in [`Approvals`] should exist in [`Proposals`]. Note, that this
+	///    automatically implies Approvals.count() <= Proposals.count().
+	/// 5. [`SpendCount`] >= [`Spends`].count(). Follows the same reasoning as (1).
+	/// 6. For each spend entry contained in `Spends` we should have spend.expire_at > spend.valid_from.
+	#[cfg(any(feature = "try-runtime", test))]
+	fn do_try_state() -> Result<(), sp_runtime::TryRuntimeError> {
+		ensure!(
+			ProposalCount::<T, I>::get() as usize >= Proposals::<T, I>::iter().count(),
+			"Actual number of proposals exceeds `ProposalCount`."
+		);
+
+		let (total_amount_bonded, non_zero_count) = Proposals::<T, I>::iter().fold(
+			(BalanceOf::<T, I>::zero(), 0usize),
+			|(sum, count), (_index, proposal)| {
+				if proposal.bond != BalanceOf::<T, I>::zero() {
+					(sum + proposal.bond, count + 1)
+				} else {
+					(sum, count)
+				}
+			},
+		);
+		ensure!(
+			T::ProposalBondMinimum::get() * non_zero_count.saturated_into() <= total_amount_bonded,
+			"Total amount bonded falls short of required minimum."
+		);
+		if let Some(proposal_bond_maximum) = T::ProposalBondMaximum::get() {
+			ensure!(
+				proposal_bond_maximum * non_zero_count.saturated_into() >= total_amount_bonded,
+				"Total amount bonded exceeds required maximum."
+			);
+		}
+
+		Approvals::<T, I>::get()
+			.iter()
+			.try_for_each(|proposal_index| -> DispatchResult {
+				ensure!(
+					Proposals::<T, I>::contains_key(proposal_index),
+					"Proposal indices in `Approvals` must also be contained in `Proposals`."
+				);
+				Ok(())
+			})?;
+
+		ensure!(
+			SpendCount::<T, I>::get() as usize >= Spends::<T, I>::iter().count(),
```

Comment:
I think the value of `SpendCount` is more interesting in the fact, that it should describe the next free index in the `Spends` map.  
So you could check that all indices in the `Spend` map are strictly smaller than this.

Reply:
@ggwpez: I have added a check for this one and the equivalent one for `Spend`, but I kept the original check as well, since these are somewhat independent, I think. Depending on how much you trust the code that is.